### PR TITLE
Support for axis parameter in linalg.gemm

### PR DESCRIPTION
**Pull request by asmushetzel**
## Description ##
This PR adds an optional axis parameter to the linalg.gemm/linalg.gemm2 operators that specifies the axis that indexes the matrix rows. Default is axis = -2 which is the behavior so far for this operators (matrices are encoded by the last two dimensions). 
The rationale behind this PR is that in some important uses cases (example is the attention mechanism in the transformer model for neural machine translation) situations requiring a batched matrix-matrix multiply with a non-standard axis for the matrix rows arise naturally. Such computations can be always be performed by an explicit swap-axis/transpose operator followed by a batch-dot, but this adds significant computation overhead. As the underlying blas-libraries are able to deal natively with non-consecutive matrix representations w/out performance impact,  it is useful to leverage this and expose a higher level switch such that such additional transpositions can be omitted.

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ X] Changes are complete (i.e. I finished coding on this PR)
- [ X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [X ] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [ X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### CODE REVIEW

**Comment by asmushetzel** [comment]
One check failed on one node and this is something completely unrelated :-(

**Comment by piiswrong** [comment]
please rebase and try again

**Comment by piiswrong** [comment]
This this behavior for axis common for other frameworks?

**Comment by asmushetzel** [comment]
Other frameworks apparently do not offer this additional flexibility. You would have to go through explicit transpose/swap-axis like functions.

**Comment by asmushetzel** [comment]
The build system is currently unstable and produces random failures :-(

File: src/operator/tensor/la_op.h [hunk 1]
```diff
@@ -53,13 +54,17 @@ struct LaMatrixMacParam : public dmlc::Parameter<LaMatrixMacParam> {
     DMLC_DECLARE_FIELD(beta)
       .set_default(1.0)
       .describe("Scalar factor multiplied with C.");
+    DMLC_DECLARE_FIELD(axis)
+      .set_default(-2)
+      .describe("Axis corresponding to the matrix rows.");
```
[the remark this answers is not in the document]
Comment (asmushetzel) [reply src/operator/tensor/la_op.h]: It's not similar to the theano's tensordot

**Comment by asmushetzel** [comment]
It does solve a different use case. linalg.gemm deals with a batch of gemm operations. The extension in this PR relaxes the constraint to have only the leading dimension used as a batch coordinate (by allowing that the coordinate associated with the matrix rows is at a different axis).
 
Batching and tensordot are two different concepts. That is why Theano also has a batched_tensordot() operator. While a single matrix-matrix product can be formulated either as gemm or as a tensordot, a batch of gemms can not be formulated as a single tensordot. Suppose we have a batch-gemm on shapes A= (I, M, K) and B = (I, K, N) where the first coordinate is the batch coordinate, then batch_gemm will return a shape (I, M, N) while tensordot(A, B, [[2], [1]]) will return a shape (I, I, M, N), i.e. it is a completely different computation.   
So this PR allows more flexibility about which axis are batch coordinates, but it is not anything that brings in any tensordot-functionalities.

**Comment by piiswrong** [comment]
Ok. Thanks for the explanation.

**Pull request merged**